jina/serve/runtimes/worker/batch_queue.py

```diff
@@ -0,0 +1,187 @@
+from jina.serve.executors import BaseExecutor
+from jina.types.request.data import DataRequest
+
+# TODO: Make sure we dont deserialize the request twice
+# TODO: SIGINT/SIGTEM: Flush
+
+async def sleep_then_set(time_seconds: int, event: Event):
+    """Sleep for time_seconds and then set the event"""
+    asyncio.sleep(time_seconds)
+    event.set()
+
+class BatchQueue():
+    """A batch queue that holds the data request and the executor."""
+
+    def __init__(self, executor: BaseExecutor, exec_endpoint: str, args: Any, preferred_batch_size: int=4, timeout: int=10_000) -> None:
+        self._preferred_batch_size: int = preferred_batch_size
+        self._timeout: int = timeout
+        self._executor: BaseExecutor = executor
+        self._exec_endpoint: str = exec_endpoint
+        self.args = args
+
+        self._reset()
+    
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__}({self._executor}, {self._exec_endpoint}, preferred_batch_size={self._preferred_batch_size}, timeout={self._timeout})'
+    
+    def __str__(self) -> str:
+        self.__repr__()
+    
+    def _reset(self) -> None:
+        """Set all events and reset the batch queue."""
+        if getattr(self, '_events', None) is not None:
+            [i.set() for i in self._events]
+        self._events: List[Event] = []
+        self._requests: List[DataRequest] = []
+        self._request_lens: List[int] = []
+        self._big_doc: DocumentArray = DocumentArray.empty()
+
+        self._flush_trigger: Event = Event()
+        asyncio.create_task(self.flush(self._flush_trigger))
+        self._timer_started = False
+    
+    def _start_timer(self):
+        asyncio.create_task(sleep_then_set(self._timeout, self._flush_trigger))
+        self._timer_started = True
+    
+    def _reset_timeout(self) -> None:
+        """Reset the timeout event."""
+        self._timeout_event.set()
+        self._timeout_event = Event()
+    
+    async def push(self, request: DataRequest, event: Event) -> None:
+        """Append request to the queue. Once the request has been processed, the event will be set.
+        
+        :param request: The request to append to the queue.
+        :param event: The event that will be set once the request has been processed.
+        """
+        # TODO: What if timer ends and triggers the flush in the middle of this function? Is that possible?
+        # This function is blocking and thus should be atomic right?
+        if not self._timer_started:
+            self._start_timer()
+
+        docs = BatchQueue.get_docs_from_request(
+            [request],
+            field='docs',
+        )
+
+        self._big_doc.extend(docs)
+        self._requests.append(request)
+        self._request_lens.append(len(docs))
+        self._events.append(event)
+
+        if len(self._big_doc) > self._preferred_batch_size:
+            self._flush_trigger.set()
+    
+    async def flush(self, trigger_event: Event) -> None:
```

Comment:
this doesn't flush immediatelly, but rathe it awaits the event and then flushes. So maybe call it `await_and_flush()` or something like that?